src/LmdbCache.ts

```diff
@@ -30,113 +34,153 @@ export class LmdbCache<V = any> implements SortKeyCache<V> {
-      start: `${key}|${sortKey}`,
+      start: `${key}${this.subLevelSeparator}${sortKey}`,
       reverse: true,
       limit: 1
     }).asArray;
-    if (result.length) {
-      if (!result[0].key.startsWith(key)) {
-        return null;
-      }
-      return {
-        sortKey: result[0].key.split('|')[1],
-        cachedValue: result[0].value
-      };
-    } else {
-      return null;
+
+    if (result.length && result[0].key.startsWith(key)) {
+      return this.joinedKeyResultToSortKeyCache(result[0]);
+    }
+    return null;
+  }
+
+  private async joinedKeyResultToSortKeyCache(joinedKeyResult: {
+    key: string;
+    value: ClientValueWrapper<V>;
+  }): Promise<SortKeyCacheResult<V> | null> {
+    const wrappedValue = joinedKeyResult.value;
+
+    if (wrappedValue && wrappedValue.tomb === undefined && wrappedValue.value === undefined) {
+      return new SortKeyCacheResult<V>(joinedKeyResult.key.split(this.subLevelSeparator)[1], wrappedValue as V);
     }
+
+    if (wrappedValue && wrappedValue.tomb === false && wrappedValue.value != null) {
+      return new SortKeyCacheResult<V>(joinedKeyResult.key.split(this.subLevelSeparator)[1], wrappedValue.value);
+    }
+    return null;
   }
 
   async put(cacheKey: CacheKey, value: V): Promise<void> {
     return this.db.childTransaction(() => {
-      this.db.put(`${cacheKey.key}|${cacheKey.sortKey}`, value);
-
-      // Get number of elements that is already in cache.
-      // +1 to account for the element we just put and will be inserted with this transaction
-      const numInCache =
-        1 +
-        this.db.getKeysCount({
-          start: `${cacheKey.key}|${genesisSortKey}`,
-          end: `${cacheKey.key}|${cacheKey.sortKey}`
-        });
+      this.doPut(cacheKey, new ClientValueWrapper(value));
+      this.removeOldestEntries(cacheKey);
+    });
+  }
 
-      // Make sure there isn't too many entries for one contract
-      if (numInCache <= this.lmdbOptions.maxEntriesPerContract) {
-        // We're below the limit, finish
-        return;
-      }
+  async del(cacheKey: CacheKey): Promise<void> {
+    await this.doPut(cacheKey, new ClientValueWrapper(null, true));
+  }
 
-      // Remove the oldest entries, so after the final put there's minEntriesPerContract present
-      const numToRemove = numInCache - this.lmdbOptions.minEntriesPerContract;
-      // Remove entries one by one, it's in a transaction so changes will be applied all at once
-      this.db
-        .getKeys({
-          start: `${cacheKey.key}|${genesisSortKey}`,
-          limit: numToRemove
-        })
-        .forEach((key) => {
-          this.db.remove(key);
-        });
-    });
+  private async doPut(cacheKey: CacheKey, value: ClientValueWrapper<V>): Promise<boolean> {
+    const putResult = await this.db.put(this.dbEntryKey(cacheKey), value);
+    if (putResult) {
+      const previousCalls = this.rollbackBatch;
+      const db = this.db;
+      this.rollbackBatch = () => {
```

Comment:
I'm not sure I like this solution...this.rollbackBatch is class field (for an instance that can be considered a 'signleton' in a warp instance context), isn't there a risk that during contract evaluation (if we're reading/writing states to/from other contracts) - we will overwrite this value before the previous one was effectively used?

Reply:
re. 'singleton' - I'm talking about the json-based state - where all contracts' state is stored in a single db file and instance.